src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ObjectWriter/Dwarf/DwarfBuilder.cs

```diff
@@ -0,0 +1,485 @@
+        public TargetArchitecture TargetArchitecture => _architecture;
+        public byte TargetPointerSize => _targetPointerSize;
+        public int FrameRegister => _frameRegister;
+
+        public uint ResolveOffset(uint typeIndex) => typeIndex == 0 ? 0u : _dwarfTypeOffsets[typeIndex - 1];
+
+        public void Write(
+            SectionWriter infoSectionWriter,
+            SectionWriter stringSectionWriter,
+            SectionWriter abbrevSectionWriter,
+            SectionWriter locSectionWriter,
+            SectionWriter rangeSectionWriter,
+            SectionWriter lineSectionWriter,
+            SectionWriter arangeSectionWriter,
+            ResolveStaticVariable resolveStaticVariable)
+        {
+            WriteInfoTable(
+                infoSectionWriter,
+                stringSectionWriter,
+                abbrevSectionWriter,
+                locSectionWriter,
+                rangeSectionWriter,
+                resolveStaticVariable);
+            WriteLineInfoTable(lineSectionWriter);
+            WriteAddressRangeTable(arangeSectionWriter);
+        }
+
+        public void WriteInfoTable(
+            SectionWriter infoSectionWriter,
+            SectionWriter stringSectionWriter,
+            SectionWriter abbrevSectionWriter,
+            SectionWriter locSectionWriter,
+            SectionWriter rangeSectionWriter,
+            ResolveStaticVariable resolveStaticVariable)
+        {
+            // Length
+            byte[] sizeBuffer = new byte[sizeof(uint)];
+            infoSectionWriter.EmitData(sizeBuffer);
+            // Version
+            infoSectionWriter.WriteLittleEndian<ushort>((ushort)(_useDwarf5 ? 5u : 4u));
+            if (_useDwarf5)
+            {
+                // Unit type, Address Size
+                infoSectionWriter.Write([DW_UT_compile, _targetPointerSize]);
+                // Abbrev offset
+                infoSectionWriter.EmitSymbolReference(RelocType.IMAGE_REL_BASED_HIGHLOW, ".debug_abbrev", 0);
+            }
+            else
+            {
+                // Abbrev offset
+                infoSectionWriter.EmitSymbolReference(RelocType.IMAGE_REL_BASED_HIGHLOW, ".debug_abbrev", 0);
+                // Address Size
+                infoSectionWriter.Write([_targetPointerSize]);
+            }
+
+            using (DwarfInfoWriter dwarfInfoWriter = new(
+                infoSectionWriter,
+                stringSectionWriter,
+                abbrevSectionWriter,
+                locSectionWriter,
+                rangeSectionWriter,
+                this,
+                _codeRelocType))
+            {
+                dwarfInfoWriter.WriteStartDIE(DwarfAbbrev.CompileUnit);
+
+                // DW_AT_producer
+                dwarfInfoWriter.WriteStringReference("NetRuntime");
+                // DW_AT_language
+                dwarfInfoWriter.WriteUInt16(DW_LANG_C_plus_plus);
+                // DW_AT_name
+                dwarfInfoWriter.WriteStringReference("il.cpp");
```

Comment:
a nit:

This name matches the one used in [native objwriter](https://github.com/dotnet/llvm-project/blob/e80993c0649d5e949eb4ac23fe22290b4ecb7e66/llvm/tools/objwriter/debugInfo/dwarf/dwarfGen.cpp#L880). Could we use a description in managed implementation instead of non-existent filename; such as,  `$".NET NativeAOT generated compile unit for '{outputFilename}'"` (by propagating `outputFilename=Path.GetFileName(objectFilePath)` from ObjectWriter to DwarfBuilder)?

According to the z/OS [docs](https://www.ibm.com/docs/en/zos/2.1.0?topic=entries-source-file-name):
> If the source file is compiler generated, the name can be used to provides a description of the compiler generated file, and not necessarily a physical file name. 

which matches the scenario at hand.

e.g. dwarfdump output will read:
`Verifying unit: 1 / 116, ".NET NativeAOT generated compile unit for 'MyApp.o'"`
as opposed to:
`Verifying unit: 1 / 116, "il.cpp"`

Reply:
This is extremely finicky bit, unfortunately. The main reason the attribute is even generated is to make Apple ld64 linker happy, and that linker expects valid file name / path in the attributes.